CRM-20555: Incorrect balance amount is shown on participant view page if 'Deferred Revenue' is enabled

monishdeb opened the pull request:
* [CRM-20037: Change Contribution tab to show balance due](https://issues.civicrm.org/jira/browse/CRM-20037)

---

 * [CRM-20555:  Incorrect balance amount is shown on participant view page if 'Deferred Revenue' is enabled](https://issues.civicrm.org/jira/browse/CRM-20555)

monishdeb commented:
Here's the screenshot that shows the new column ```Balance Due``` and removed ```Premium``` under Contribution Tab
![image](https://cloud.githubusercontent.com/assets/3735621/25266004/3fef9850-268d-11e7-955b-a2322471733a.png)

pradpnayak commented:
Tested PR, result in attached SS
![test](https://cloud.githubusercontent.com/assets/2053075/25298540/64a95d6c-2713-11e7-84dc-ac83cdaeac2e.png)

pradpnayak commented:
However when i enable 'Deferred Revenue' and Purchase a General Membership with completed Contribution status i get 
![deferred](https://cloud.githubusercontent.com/assets/2053075/25298646/778ec2cc-2714-11e7-82e7-d72f434f67c1.png)

Balance amount seems to be wrong

eileenmcnaughton commented:
We'd better get some stats on this on large searches on large sites before merging - we did discuss this on another ticket - ie. the concerns about imposing possibly-unconformant change - I'll try to find the ticket

seamuslee001 commented:
@eileenmcnaughton I think that might have been the soft credits one?

JoeMurray commented on CRM/Contribute/Selector/Search.php:
```diff
@@ -402,6 +402,13 @@ public function &getRows($action, $offset, $rowCount, $sort, $output = NULL) {
       }
 
+      $row['balance_due'] = CRM_Core_BAO_FinancialTrxn::getPartialPaymentWithType(
+        $result->contribution_id,
```
@eileenmcnaughton @colemanw and I just looked through this function and it does two calls to the db which don't look bad. This .tpl is used on contributions tab and also on search results for Search: Find Contribution. But in the latter it's only going to be perhaps 100 or so records. So we're not too concerned after code review.

eileenmcnaughton replied:
So it's a max 300 extra queries per search, or 100 viewing a contribution tab on a donor with 30 or donations?

monishdeb replied:
On further research, if we MUST add this improvement i.e. add ```Balance Due``` column on Contribution list, then there is no other easy alternative other than using ```CRM_Core_BAO_FinancialTrxn::getPartialPaymentWithType(..)``` to calculate the balance due amount.

eileenmcnaughton replied:
OK - it would actually be easy to do in an extension - ie. add an api in the extension & just add a region in the tpl & call crmAPI from there

JoeMurray replied:
So @eileenmcnaughton do you have site to test this on, or are you proposing that JMA create a test site and populate it. Personally I don't think this is an issue. Let's see what the total number of queries per search results page is for 25 results with and without the change, and the differential page load time on a server with reasonable resources, and then discuss with @colemanw ? Recall as well that we are on a deadline on this and we had agreement to move ahead.

eileenmcnaughton replied:
Hi - if you have agreed to go ahead with @colemanw then I guess he'll have to honour that - please do it as a setting that is opt-out by default and then no-one else will be affected.

If you DO impose this on all sites then you effectively transfer the work onto me of making it not run extra queries before I can upgrade.

monishdeb commented:
For https://github.com/civicrm/civicrm-core/pull/10204#issuecomment-296324291 

@pradpnayak on closer look ```CRM_Core_BAO_FinancialTrxn::getPartialPaymentWithType(...)``` doesn't return proper ```balance due``` amount if 'Deferred Revenue' is enabled because it is not handled in that BAO function. There's another place where this BAO function is used and that is participant's 'Fee Section' page where it show the same incorrect 'Balance'. So I think we need to raise a separate ticket about adding support for ```Deferred Revenue``` in ```CRM_Core_BAO_FinancialTrxn::getPartialPaymentWithType(...)``` and that will solve the issue in both the places.

@JoeMurray should I create a separate ticket for that?

monishdeb commented:
@pradpnayak I have fixed the issue where it shows the incorrect balance amount if 'Deferred Revenue' is enabled (also added unit test) and this is the patch https://github.com/civicrm/civicrm-core/pull/10204/commits/52d2e894bf23678c0378355e06bbfe3955790845 

This has also fixed the issue on Fee Section on Participant's view page. Here's the difference: 
1. Before the fix (check the balance and total paid amount)
![image](https://cloud.githubusercontent.com/assets/3735621/25378652/4c0d4004-29c8-11e7-879b-299f6eb51b7c.png)

2. After the fix 
![image](https://cloud.githubusercontent.com/assets/3735621/25378671/5aeaa13e-29c8-11e7-9354-aa08a42af9a2.png)

monishdeb commented:
@JoeMurray @eileenmcnaughton I have made necessary changes in this PR to add ```Balance Due``` column using searchColumns hook. There is a slight change in the position of this new column, which is now placed next to ```Actions``` column due to limitation caused by underlying code structure. Here's the screenshot: 
Screen Shot 2017-05-03 at 2.45.32 PM
 And as per the core changes here's my hook definition 
```
function hook_civicrm_searchColumns($contextName, &$columnHeaders, &$rows, $form) {
  if ($contextName == 'contribution') {
    $columnHeaders['hookHeaders'] = array(
      'balance_due' => array(
        'name' => ts('Balance Due'),
      ),
    );

    foreach ($rows as $key => $row) {
      $balanceDue = CRM_Core_BAO_FinancialTrxn::getPartialPaymentWithType(
        $row['contribution_id'],
        'contribution',
        FALSE,
        $row['total_amount']
      );
      $rows[$key]['balance_due'] = sprintf("<b>%s</b>", CRM_Utils_Money::format($balanceDue));
    }
  }
}
```

eileenmcnaughton commented:
#10295 proposes an alternative approach to the tpl

monishdeb commented:
@eileenmcnaughton I have reverted my changes in tpl and php files after merging https://github.com/civicrm/civicrm-core/pull/10295 Also please check my [comment](https://github.com/civicrm/civicrm-core/pull/10295#issuecomment-299167251)

eileenmcnaughton commented:
@monishdeb  this seems OK except - it's not fixing the thing the ticket is about anymore, it's fixing something else related to deferred revenue. Can you log a separate ticket that describes what this is fixing & update the PR /commits to refer to it.

eileenmcnaughton commented:
also - you've updated the docs for how to do the hook?

monishdeb commented:
@eileenmcnaughton I have created a new ticket and updated the title of the PR. Also submitted a minor fix https://github.com/civicrm/civicrm-core/pull/10334

eileenmcnaughton commented:
Getting the JIRA admin sorted was the final issue on this PR - adding merge on pass

monishdeb commented:
Jenkin, test this please
